### DISCUSSION TITLE
(Yet another) git@github.com: Permission denied (publickey)

### DISCUSSION THREAD

**Discussion by StyXman** [community/community#39107] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I just created a ssh key pair that is specific for only one clone. I added the priv part to my ssh agent and the public to my GH profile. I can `ssh` just fine:

```
22:05 $ ssh -T git@github.com
Hi StyXman! You've successfully authenticated, but GitHub does not provide shell access.
```

But I can't pull or push from this clone:

```
22:05 $ git push
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
```

This is the remote config:

```
22:07 $ git remote -v
origin  git@github.com:StyXman/elevation.git (fetch)
origin  git@github.com:StyXman/elevation.git (push)
```

Any ideas?

**Comment by david-engelmann** [comment]
[Check out this issue on ssh troubleshooting, it's possible your config isn't pointing to the correct url](https://github.com/tpope/vim-fugitive/issues/2074)

**Reply by StyXman** [reply]
Yes, I tried them all and yes, it's the right URL.

**Comment by airtower-luna** [comment, marked as answer]
That looks like Git and SSH don't agree about where your SSH keys should be. Do you happen to be on Windows? I've heard about such things happening there before. :sweat_smile: 

You can set something like [`GIT_SSH_COMMAND="ssh -v"`](https://git-scm.com/docs/git#Documentation/git.txt-codeGITSSHCOMMANDcode) to see what SSH does when called from Git.

**Reply by StyXman** [reply]
That is news to me :) No, this is Linux, and now, for some reason, it's working. Maybe the push/pull system is another one and hadn't sync'ed the key then? I was trying for 10s of minutes...

**Reply by airtower-luna** [reply]
Maybe `SSH_AUTH_SOCK` wasn't set so it'd find your SSH agent, at least that'd explain why it didn't work even though you had added the key to the agent. :thinking: 

Anyway, I'm glad you got it working!

**Reply by StyXman** [reply]
The agent was being found, as `ssh -T git@github.com` _was_ working, and I didn't touch the envvar. Just waiting overnight seems to have fixed it (???).

**Comment by StyXman** [comment]
Can I close this one?

**Comment by pailgrape** [comment]
> ### Select Topic Area
> Question
> 
> ### Body
> I just created a ssh key pair that is specific for only one clone. I added the priv part to my ssh agent and the public to my GH profile. I can `ssh` just fine:
> 
> ```
> 22:05 $ ssh -T git@github.com
> Hi StyXman! You've successfully authenticated, but GitHub does not provide shell access.
> ```
> 
> But I can't pull or push from this clone:
> 
> ```
> 22:05 $ git push
> git@github.com: Permission denied (publickey).
> fatal: Could not read from remote repository.
> ```
> 
> This is the remote config:
> 
> ```
> 22:07 $ git remote -v
> origin  git@github.com:StyXman/elevation.git (fetch)
> origin  git@github.com:StyXman/elevation.git (push)
> ```
> 
> Any ideas?

I'm new here. not sure if it matters that i comment since it's answered, but i have a bit of experience. 

when using sigs, I'd have issue with auth when they are just created too.  i found that by waiting for a little time to pass would solve it. basically, what happens, is i found that my OS needed time to negotiate a trust. i found that using a TPM worked best, but found it difficult to actually use. fTPM seem to be slower in certain systems and environments are very important. the least of my problems was actually using it when I needed to was difficult. power and resources were a factor not always looked after. securing and safe guards where useless during maintenance. best approach was offline.

**Reply by airtower-luna** [reply]
> found that my OS needed time to negotiate a trust.

As you can read above I can't say what exactly caused the problem here, but this is definitely unrelated. When using Git over SSH you create your key pair (that may take some time) and then upload the public key to the server (here: GitHub), *and tell the server to trust it*. From that point the server will trust that key for your login, no further negotiation needed. :slightly_smiling_face: 

Technically it'd be possible that some hiccup in the infrastructure causes a delay, I suppose, but then connecting directly with `ssh` instead of `git` wouldn't have worked. :sweat_smile:

**Reply by pailgrape** [reply]
funny! how?  it's not funny waiting for secure connections. it's always been a waiting game. the speedier the network the quicker this is. I have no git software myself, so no sure what's happening. i know from an unbuntu NAS that the keys would sometimes not work with win7 when the pc snoozed. i would have to reset the password for the NAS account, and re-auth win credentials, and re link any share.. that's not funny. so actually very likely that it's a server side cred etc. and update slowing it down.

**Comment by bondarenkod** [comment]
@StyXman  try this https://gist.github.com/bondarenkod/20e96267a46f78bf98d3e999d04c102f
